### ISSUE TITLE
False positive with `typing.Self` return type and `@typing.final`

### ISSUE THREAD

**Issue by runfalk** [microsoft/pyright#6746]
I just updated from `1.1.334` to `1.1.340` and got a bunch of type checker errors related to the use of `typing.Self` within instance methods that returns a copy of the same type. I can see how this could be a problem if that class is ever extended, but even if I mark it as `@typing.final` (https://docs.python.org/3/library/typing.html?highlight=final#typing.final).

```python
import typing as t

@t.final
class SelfFromClsMethodFinal:
    @classmethod
    def create(cls) -> t.Self:
        return cls()

    @classmethod
    def create_with_name(cls) -> t.Self:
        # Expression of type "SelfFromClsMethodFinal" cannot be assigned to
        # return type "Self@SelfFromClsMethodFinal"
        return SelfFromClsMethodFinal()

    def copy(self) -> t.Self:
        return type(self)()

    def copy_with_name(self) -> t.Self:
        # Expression of type "SelfFromClsMethodFinal" cannot be assigned to
        # return type "Self@SelfFromClsMethodFinal"
        return SelfFromClsMethodFinal()
```

I believe marking it as `final` should allow type checking to pass as there is no way the return type would not be `Self`?

**Comment by erictraut** [comment]
Pyright's behavior here is correct. This isn't a bug.

Making the class `final` doesn't change the fact that you're using `Self` incorrectly here.

If your method is returning an instance of `SelfFromClsMethodFinal`, then its return type should indicate as such. In that case, `Self` is not the right annotation to use.

**Comment by mikeshardmind** [comment]
> If your method is returning an instance of SelfFromClsMethodFinal, then its return type should indicate as such

That's what `Self` is in this context. And it can't be anything else either. The definition of Self as a type is not one of the underspecified things. [There's also a direct example with classmethods in the PEP](https://peps.python.org/pep-0673/#use-in-classmethod-signatures)
